course5/homework2/carmack/carmack.cc

```diff
@@ -0,0 +1,39 @@
+#include <iostream>
+#include <math.h>
+#include <time.h>
+#include <string>
+
+// Implementing fast-sqrt algortihm. However, for me, sqrt from the math library seems much faster.
+
+float inv_sqrt( float x )
+{
+    int i;
+    float x2 = 0.5 * (x);
+    float y = x;
+    const int magic = 0x5f3759df;
+
+    float* p = &y;
+    int* p2 = reinterpret_cast<int*>(p);
+    i = *p2;
+    i = * reinterpret_cast<int*>(&y);
+    i = magic - ( i >> 1);
+
+    y = *reinterpret_cast<float*>(&i);
+    y = y * ( 1.5 - ( x2 *y *y) );
+
+    return y;
+}
+
+int main(int argc, char** argv)
+{
+    float x = std::stof(argv[1]);
+    clock_t t;
+    t = clock();
+    std::cout << "Carmack_sqrt of " << argv[1] << "=" << 1/inv_sqrt(x) << std::endl;
+    t = clock() - t;
+    std::cout << "Carmack_time" << "=" << t << std::endl;
+    t = clock();
+    std::cout << "math_sqrt of"  << "=" << sqrt(x) << std::endl;
+    t = clock() - t;
```

Comment:
It's a good idea to benchmark these two implementations against each other, but not this way.
In this case the time of writing to output will overshadow the time of calculating the square root.
I'll show you how to do it properly one day.